modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py

```diff
@@ -71,7 +92,12 @@ def __init__(self, data, length=None, width=None, ip=None, call_queue=None):
             )
         )
 
-    def apply(self, func, *args, **kwargs):
+    def __del__(self):
+        """Decrement the reference counter."""
+        if isinstance(self._data_ref, DeferredExecution):
+            self._data_ref.ref_count(-1)
```

Comment:
@AndreyPavlenko thank you for addressing many of my comments. I guess I have one last question regarding reference counting.

At first glance, it seems that since there is a `ref_count(-1)` in the destructor, then there should be a symmetrical `ref_count(1)` in the partition constructor. In this case, we can also remove `ref_count(1)` from the `__copy__` function.

Reply:
This is how it was before, but now `ref_count(1)` has been moved to the `DeferredExecution()` constructor. We can't have `ref_count(1)` in both constructors.

Reply:
We can decrement the reference when passing DeferredExecution to the partition constructor.

Reply:
```python
      de = DeferredExecution(self._data_ref, func, args, kwargs)
        if LazyExecution.get():
            de.ref_count(-1)
            return self.__constructor__(de)
```

Reply:
It's very similar to what it is now.

My main motivation is to reduce the number of potential problems associated with references counting. To do this, we need to narrow the area of interaction, and preferably without delaying it for later. @YarShev @dchigarev maybe you have any good suggestions?

Reply:
One of the solutions is to use weak references to the partitions, but it may have a negative performance impact.